### ISSUE TITLE
Playtesters wanted for new SH2 mouse support

### ISSUE THREAD

**Issue by Polymega** [elishacloud/Silent-Hill-2-Enhancements#618]
@MasterDarkseid @MarioTainaka @DaniloElectricMonkeys @Veynam @Sumea @Balthazor44 @PedroC47 @Gwektro @shpatter @vStar925

(Tagging you all since you playtested for us previously and did a great job.)

It's with great pleasure that I announce full mouse support is coming to the game through our project! Thanks to the tireless efforts and excellent work by @mercury501, this new feature will give you an in-game mouse experience similar to what's featured in SH3 PC. Now, with the mouse, you can:

- Turn James by moving the mouse left/right
- Select/examine things with left click
- Ready your weapon with right click during normal gameplay
- Back out of menus/screens/text with right click
- Cycle weapons forward with mouse wheel up
- Cycle weapons down with mouse wheel down

In addition to this, Mercury has added mouse support for the following screens:

- Mouse support for navigating the pause screen menu
- Mouse support for navigating the memos list screen

^ While in these screens, you can move the mouse to cycle through their selections!

## What to test for
Play the game using the mouse and make sure everything is working well and working right. As you play, just make sure nothing odd is going on as you use the mouse. We recommend you take your time playing. 

Try and perform every conceivable action you can think of with the mouse as you play. Try and mash mouse clicks together during certain scenes or text prompts; time your clicks in weird ways to see if it breaks anything; Alt + Tab out of the game while playing to make sure everything is fine there; do anything you can think to try and break things.

## Prerequisites to test
You will need to use the test build (linked below) for testing. Also, having experience using the mouse with SH3 PC will help you out here.

- Download the test build here: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10103731/d3d8.zip) **(Updated 2022-11-28 07:33 AM EST)**
- Mouse support for the functions below require that they are assigned to a keyboard input in order to work. Make sure the following functions are assigned keyboard inputs in the game's Options > Control Options screen:
   - Action/Confirm
   - Ready Weapon
   - Cancel
   - Next Weapon
   - Previous Weapon
- It is strongly recommended to use a keyboard layout similar to the diagram below for the best playing experience while using a mouse:

![image](https://user-images.githubusercontent.com/27395260/195458651-4e66ab8e-1c11-481e-ae79-9d8ac83ea70a.png)

# Thank you for your help!
We want this new feature to be as solid and stable as possible upon public release, so we appreciate your help in testing! Once fully tested, please leave your notes in this thread. Let us know if you run into any problems, but also let us know if everything went smoothly for you; we'll need to know either way!

**Comment by Polymega** [comment]
Oh! A few follow-up notes:

- Mouse support only works with 3D (rotational) control movement. It will be disabled for 2D (directional) movement. 
   - This is by design and the same happens in SH3 PC. You cannot control the character using a mouse in 2D mode.
- Strafing does not work in 2D mode. T his has always been the case for SH2. 
- We are looking into a way to have Strafe keys automatically change to Turn keys if switched to 2D mode (for convenience) or while in menus/examining puzzles (for practicality). 
   - This would also allow the Strafe keys to be used for menu/puzzle navigation, which would be especially useful when assigning Strafe to the A & D keys.

**Comment by Balthazor44** [comment]
Hello! I would be delighted to do testing on this. I should be able to begin later tonight.

**Comment by Balthazor44** [comment]
I noticed something a bit unusual just now. I am examining the "Patient Memo" in the hospital office. The right-click functions differently when reading this memo there in the room vs. reading it in the inventory. In the inventory, the right-click works the same as the left-click. But when reading the memo in the room, the right-click essentially skips a page because it moves to the next page on both the button down and the button up. I see a similar thing happening in other places, like the drain pipe in the hospital. Left-click on the drain pipe to trigger the text, then right-click and it skips ahead two pages. Also the diary on the hospital roof.

Other than that, everything has been working great. I've been running around going click crazy on everything, I've Alt-Tabbed many times, no problems yet.

Using the mouse wheel to cycle through weapons is handy, but interestingly it reveals how James' stance changes slightly with each equipped weapon. It makes sense that it would, it's just something I never noticed before. But that makes it so that when you spin the mouse wheel quickly it makes it look like James is having a seizure or something. If you hit a strafe button or run while spinning the mouse wheel, then things get really funky.

**Comment by mercury501** [comment]
Hi @Balthazor44 , thank you for your time and enthusiasm! I'll test this as soon as I get home.

In the build is included a debug overlay, you can bring it up by pressing `CTRL + G`, and it displays the current mouse function, aim or Cancel, it could be useful for testers when reporting bugs like this one!

**Comment by Polymega** [comment]
@Balthazor44, thanks for helping us test! **@ALL**: Please also consider helping us test!

So Balthazor, I believe this is an issue with consistency in the game itself. For some memos/puzzles, the developers decided that pressing the Back button can advanced the text all the way through while other times you must press Confirm to get things started.

I mostly noticed this happening in the Hospital area which, as it's often agreed upon, was one of the first levels/environment made by the development team. Perhaps they didn't agree on a "rule" for button pressing for full screen images at this time and, upon agreeing with a set of rules later on, forgot to go back and apply those rules to these earlier screens.

@mercury501 
> edit2- or I can "hide" the rmb from the game when EventIndex = In game. Shouldn't be a problem, as when rmb is used by the game is in map, menus or inventory.

Do you mean "hide" the Cancel input for RMB while playing? If so, this is needed to back out of a memo/puzzle while playing.

**Comment by mercury501** [comment]
I mean to completely hide the rmb input from the game, you would back out using our injected cancel input

**Comment by MasterDarkseid** [comment]
I've just tried a few minutes before going to bed but i can't turn right and left using the mouse is there anything to configure.

Like said above.

However everything else work.

**Comment by mercury501** [comment]
> I've just tried a few minutes before going to bed but i can't turn right and left using the mouse is there anything to configure.
> 
> Like said above.
> 
> However everything else work.

Hi @MasterDarkseid , you must have EnableEnhancedMouse = 1 and MemoScreenFix = 1 in your d3d8.ini!

**Comment by mercury501** [comment]
@Balthazor44 , try this build if you like, * should * fix the issues you've encountered.
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9779549/d3d8.zip)

**Comment by Balthazor44** [comment]
> @Balthazor44 , try this build if you like, * should * fix the issues you've encountered. [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9779558/d3d8.zip) edit- wrong build

Ok great, I will give that a whirl tonight.

**Comment by Polymega** [comment]
Whenever right click changes to cancel (such as when text is displayed or viewing a full screen image), if you hold down right click does the cancel signal get sent every frame? Here are examples of holding right click down:

https://user-images.githubusercontent.com/27395260/195721277-6d1346e6-e7ef-42fa-8792-36cc75166e64.mp4

https://user-images.githubusercontent.com/27395260/195721286-e6d2372f-bffb-46eb-8aba-3521c21c95ad.mp4

This behavior is happening on the latest build and the one before it. I tried holding down Esc and P (set to "Pause") on the keyboard, but didn't experience this behavior. "P" didn't do anything at all, and for Esc, I had to press it each time to progress the text.

**Comment by Balthazor44** [comment]
I tried out the latest build on a couple of the documents and it seems to work great. I'm also not getting the same behavior holding the right click down that Polymega described above, when I hold down the right click it advances the page on the down click and then just stays there. I'll continue doing more testing tonight.

Quick question, is there an option anywhere to change the mouse sensitivity? It seems like I have to move the mouse quite a lot when moving James around. It's not terrible or anything, but it would be nice to boost the sensitivity just a bit to see how that feels.

**Comment by Polymega** [comment]
> I'm also not getting the same behavior holding the right click down that Polymega described above, when I hold down the right click it advances the page on the down click and then just stays there.

Interesting. I closed and relaunched the game several times and got the same results that I shared. Did you test those two particular memos that I did? Maybe this is a per-memo issue?

**Comment by Balthazor44** [comment]
> Interesting. I closed and relaunched the game several times and got the same results that I shared. Did you test those two particular memos that I did? Maybe this is a per-memo issue?

I did try the garbage chute memo, but not the typewriter memo. I also tried it on the patient memo in the hospital memo too and that worked fine as well.

**Comment by Polymega** [comment]
PS -

> It seems like I have to move the mouse quite a lot when moving James around. It's not terrible or anything, but it would be nice to boost the sensitivity just a bit to see how that feels.

Let's see what Mercury says about that but, in the meantime, trying playing the game in 60 fps to see if that helps any. I feel like his turning is tighter at this framerate, but it might just be a placebo effect I'm experiencing.

**Comment by Balthazor44** [comment]
> Let's see what Mercury says about that but, in the meantime, try playing the game in 60 fps to see if that helps any. I feel like his turning is tighter at this framerate, but it might just be a placebo effect I'm experiencing.

Thanks, but I doubt this machine could handle 60 fps. Maybe I'll try that on my other one to see how that feels. But overall I hate playing this game with a mouse, so once this testing is done, I likely won't be playing it this way again. But if it's possible to change the mouse sensitivity, I'm sure that's something that folks who will play the game with a mouse would like to be able to adjust.

**Comment by Gwektro** [comment]
Hey, thanks for the invite again! I don't have a ton of time lately, but I decided to mess with the current build that was uploaded earlier today. I have a good amount of exp. using keyboard & mouse for SH3 as well as just keyboard for this one & SH4, so far this iteration of it feels like an improvement from 3. 

The sensitivity actually feels a bit higher & smoother from 3, the turn speed stops accelerating pretty quickly the moment you start moving your mouse. So the slower you move your mouse, the faster James turns too. Your hand shouldn't be moving that fast at all or else too quick of a swipe will just throttle the turn speed altogether since it only goes so fast which makes James turn slower in the end. Also framerate nor DPI makes little to no difference, I switched & tested it just to be sure. Its the same logic in SH3 as well.

In terms of 'issues', I triggered almost nothing except something pretty buggy related to bindings, but the previous build gave me what looked like some 'autoskip' behavior holding down RMB on every memo up to the end of the apartment section (what @Polymega showed in the video above), but in the latest build that was no longer the case & worked normally again. I rebooted the game a bunch & even reinstalled the .dll a few times to make sure. I actually liked it even though its not intended lol.

This is a little confusing to type so hopefully this will make some sense. I also feel like using keyboard with the PC port has always brought up really weird quirks when playing 2/3 with it, I have a thing when using the search camera on any binding, if you're walking/running & want to turn James + his head left/right **on mouse** I can't. If you're standing however, James will turn his head appropriately on mouse. BUT, sometimes James will turn only his head (but not his body) on mouse with the search cam active while im running forward after quickturning or quickloading or just altogether? I really cant figure out why it only happens sometimes but it did work occasionally when doing those two things for me anyways. I tried to rebind the movement actions to arrow keys instead of letters in case that was interfering with whatever the mouse is trying to do but that produced the same result. **On keyboard**, I absolutely have to bind search cam to almost anything other than TAB or else turning while walking (but not running) using the search cam forces me to only turn right if I want to turn James. There is a similar thing in SH3 too along with the mouse support but plays out a bit differently in that game. "Heather" can turn her head when walking/running using search cam on keyboard, but in that game there's separate key bindings for movement & camera control (up/down/left/right) so you can easily turn her character with that. With the mouse, she'll only turn her head if you're using that to turn with the search cam, but then turns & functions normally if you deactivate it afterwards. She'll also only turn right on keyboard when setting search cam to TAB there too lol.

So since SH2 doesn't have separate bindings for camera control + character movement & SH3 has no options to configure the mouse, it feels like the same inputs are fighting each other when you're in search cam mode on mouse. The only thing I can even think of as to why this might happen is something to do with how developers programmed keyboard bindings & maybe didn't test how certain actions worked on specific keys? & now that we get to use a mouse with the same functions, that might interfere with it more? No idea, I got nothing for this. 

I'd say in its current testing state, it feels pretty nice aside from all that up above. Cycling weapons (James can dance now!) & using all the other functions unrelated to movement worked great. My big hope would be to have the ability to use my mouse buttons as extra keys to rebind actions to there instead of my keyboard. I'm sure that's a whole other thing to work out though. :)

**Comment by mercury501** [comment]
Hi @Gwektro, thanks for your contribution.
What I can say about the search view, basically I just inject right analog stick inputs when search view is active, so it should behave exactly like when using a controller.
>So the slower you move your mouse, the faster James turns too.
Not sure about this, maybe you meant the faster you move...?


For @Balthazor44, I could put a "sensitivity" setting, but I'm not sure about how useful it would be, because sensitivity is used very little in all this, since the mouse is used to emulate a Left Analog Stick, so you could move the mouse right at 10cm/s or 1m/s, it would turn james at a constant rate. I could decrease the minimum input required to turn james at full speed, or make him turn more at low mouse input values, I'll post a fresh build as soon as I can.

**Comment by Polymega** [comment]
@Gwektro 
To expand a bit more on what Mercury said about search view: This is largely due to both character movement and search view sharing the same keyboard keys to move them around (as you said). To get around this, the devs programmed it to where, once James has fully stopped moving, the keys that moved him will be nulled so that these same keys will now move the camera.

If you activate search view while James is still moving, it creates this "fight" between the two movements as you described. So you'll find similar behavior with using the mouse for search view; since a computer only has one mouse, its movements too must be shared between James and the search camera. So that behavior described for the keyboard holds true for the mouse. 

The only exception to this is when using a controller. Since a controller has two joysticks, the game is programmed to separate James' movement on the left stick and search camera movement on the right stick. So with a controller, you can still freely move James will having full control of the search camera. But again, since we only have one mouse, we have to let the search camera movement take precedence over James' movement what it's activated.

I think it's the best compromise here, given the situation, and behaves similarly to SH3 PC, as you mentioned.

> Also framerate nor DPI makes little to no difference, I switched & tested it just to be sure. Its the same logic in SH3 as well.

Good to know; it was a placebo effect for me then. I think turning with the mouse if nice! I totally understand your preference to not use a mouse, Balthazor, as I was super turned off from it as well when I first played SH3 PC. After a while, I did get used to it on that game and came to like it. I'm not saying you'll like it yourself, but it may grow on you over time!

> And @Polymega try redownloading my last build, I sent the wrong one at first!

I did that last night and was consistently getting the issue that I shared videos for. Today, it's not happening. 😵 Did my mouse get stuck or something?! No clue what happened there then... but everything looks to be fine now. I will try and do more testing later for this. Thanks!

**Comment by Balthazor44** [comment]
> For @Balthazor44, I could put a "sensitivity" setting, but I'm not sure about how useful it would be, because sensitivity is used very little in all this, since the mouse is used to emulate a Left Analog Stick, so you could move the mouse right at 10cm/s or 1m/s, it would turn james at a constant rate. I could decrease the minimum input required to turn james at full speed, or make him turn more at low mouse input values, I'll post a fresh build as soon as I can. Oh and, give it a chance! Mouse input grows on you, i swear.

That makes sense. I don't know if I'd waste much time on it, it feels pretty decent as is, and mostly it's probably just me not entirely used to moving with a mouse. Normally I run this game through my tv and use an Xbox controller, I can't stand playing games at my desk. I just figured if there was an easy way to adjust the sensitivity that would be great, but if it's a bigger hassle to take on then probably not worth the effort.

**Comment by Gwektro** [comment]
> Hi @Gwektro, thanks for your contribution. What I can say about the search view, basically I just inject right analog stick inputs when search view is active, so it should behave exactly like when using a controller.
> 
> > So the slower you move your mouse, the faster James turns too.
> 
> Not sure about this, maybe you meant the faster you move...?
> 
> For @Balthazor44, I could put a "sensitivity" setting, but I'm not sure about how useful it would be, because sensitivity is used very little in all this, since the mouse is used to emulate a Left Analog Stick, so you could move the mouse right at 10cm/s or 1m/s, it would turn james at a constant rate. I could decrease the minimum input required to turn james at full speed, or make him turn more at low mouse input values, I'll post a fresh build as soon as I can. Oh and, give it a chance! Mouse input grows on you, i swear.
> 
> And @Polymega try redownloading my last build, I sent the wrong one at first!

If its emulating the right stick in that mode, then James should just be moving his head then which happened occasionally at random while mostly not doing so at all. Its really nothing major, just a weird quirk is all. & with the turn speed, you explained it pretty much better than me lol. So swiping your mouse left/right past a certain 'speed' = James only turning so fast, which means its working as intended obviously.

**Comment by MasterDarkseid** [comment]
So i noticed that while using the searching input the mouse don't work for direction. Is that intended or is it a bug ?

**Comment by Polymega** [comment]
> So i noticed that while using the searching input the mouse don't work for direction. Is that intended or is it a bug ?

This is due to inherent behavior with the game and how it handles the search camera on a keyboard. [This is discussed in some more detail here.](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/618#issuecomment-1278962626)

> Also i find the mouse sensitivity just a bit slow. Is it possible to increase it slightly.

Most likely not. The distance to move the mouse can be changed, but James will always turn at the same, constant speed. [This is discussed in some more detail here.](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/618#issuecomment-1278741514)

**Comment by MasterDarkseid** [comment]
> > So i noticed that while using the searching input the mouse don't work for direction. Is that intended or is it a bug ?
> 
> This is due to inherent behavior with the game and how it handles the search camera on a keyboard. [This is discussed in some more detail here.](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/618#issuecomment-1278962626)
> 
> > Also i find the mouse sensitivity just a bit slow. Is it possible to increase it slightly.
> 
> Most likely not. The distance to move the mouse can be changed, but James will always turn at the same, constant speed. [This is discussed in some more detail here.](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/618#issuecomment-1278741514)

It's fine for the search view only working for keyboard. It was just the camera sensitivity with mouse that i'm most curious about. But it's okay honestly i'm fine with it. 

Speedrun wise directional with controller say the best. Or not i'm not used using SH3 controls in SH2. but the strafe might do the trick. I need to do more testing.

Otherwise like i said seems pretty much a non issue at all.

**Comment by Balthazor44** [comment]
I spent a good amount of time testing on both my machines this weekend. I really had no issues at all. The only potential issue I had was a couple times I was moving forward while switching weapons with the mouse wheel and the forward button would essentially freeze, and I'd have to hit a different movement button to unfreeze it. But that only happened twice and I could never cause it to happen again. So probably nothing to be too concerned about.

A couple of observations though. First, it appears you can't use the mouse to steer the boat while moving forward. You have to use the keyboard for that. The mouse does steer the boat as long as the forward button isn't pressed. I have no idea if that's something that can be fixed, or it that's just how it is.

Second is the rotating room puzzle in the labyrinth, I could find no way to turn the block with the mouse, again I had to use the movement keys for that.

I'm hoping to get in a full beginning to end playthrough tomorrow, so I'll report back with the results of that.

**Comment by Polymega** [comment]
Great catch about the mouse not working on the boat section. 👍 

We can take a look into the labyrinth cube. I'm not sure if we can do mouse movement for that yet or not.

Oh, and semi-related: 
About the full screen images: Early on we decided that if a full screen image doesn't already utilize a mouse for its interaction then we will prevent mouse movement from affecting the puzzle altogether. The goal in doing that would be to make the "rules" for interactions with full screen images while in-game more consistent. (For example, using the mouse on the apartment clock or Blue Creek locked safe would probably be more irritating than useful.)

**Comment by Balthazor44** [comment]
I suppose then the mouse not working for the labyrinth cube is consistent with those other puzzles, so by the time a player gets to the cube puzzle they'd be used to using the keyboard for puzzles. So that one could be called a "feature" rather than a "bug". :) The boat does seem like the bigger problem.

**Comment by Balthazor44** [comment]
I noticed something else just now, if I hit the ctrl key (I have the search camera mapped there) then at any point after hit the i key to enter inventory, the status menu appears, as if I had hit ctrl+i. And it seems like the status menu will keep coming up every time I hit i to enter inventory until I hit ctrl+i to exit the status menu while on the inventory screen. But then the next time I hit ctrl for the search, it happens again.

Other than that, no further issues. I think I'm starting to get the hang of controlling with the mouse...but I still can't stand it. :)

**Comment by Polymega** [comment]
We caught the Ctrl thing the other day and Mercury has fixed it! Today, he has put in a fix for rowing the boat with a mouse. I reckon he'll post an updated test build soon for everyone here to play with. Thanks for catching that, though!

> I think I'm starting to get the hang of controlling with the mouse...but I still can't stand it. :)

I'm telling ya... it'll slowly grow on you! 🤣

**Comment by Balthazor44** [comment]
> I'm telling ya... it'll slowly grow on you! 🤣

I don't know about that, moving the mouse around, it just feels too much like exercise. :)

**Comment by mercury501** [comment]
Hi all!
thank you for your work testing this patch.
Here's a build where I've implemented a fix on the boat stage:
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9808675/d3d8.zip)

Let me know what you think!

**Comment by mercury501** [comment]
Oh and, if you still haven't tried the last one yet, try this one that should be virtually identical to the last one, save for some refactoring:
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9813964/d3d8.zip)

Thanks!

**Comment by Balthazor44** [comment]
> Oh and, if you still haven't tried the last one yet, try this one that should be virtually identical to the last one, save for some refactoring: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9813964/d3d8.zip)
> 
> Thanks!

I plan on doing some testing on the new build tonight. Does this just fix the boat issue, or does it fix the Ctrl issue as well?

**Comment by MasterDarkseid** [comment]
I'll test this new build shortly.

**Comment by Polymega** [comment]
I think there were some issues with that build, in regards to using strafing for navigating menus. I'd recommend waiting for Mercury to upload a new build.

**Comment by mercury501** [comment]
Here's a new build!
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9827142/d3d8.zip)

Keep in mind what @Polymega  said!

**Comment by MasterDarkseid** [comment]
Silent hill 2 have a huge trend right now in twitch and socials. It's good to see Silent Hill back from the dead.

**Comment by Balthazor44** [comment]
> Silent hill 2 have a huge trend right now in twitch and socials. It's good to see Silent Hill back from the dead.

Indeed. I was groaning at the possibility of a SH2 remake, but the teaser video looks pretty good, and with Masahiro Ito involved and with it looking like they're sticking pretty close to the original, there seems a chance it won't absolutely suck. But the interest generated by these recent announcements may cause more people to take a look at this Enhanced Edition as well, which would be great.

**Comment by Polymega** [comment]
Yep, there has been a boom in people talking and recommending our project since the announcement. It's exploded a bit on social platforms.

**Comment by MasterDarkseid** [comment]
Yes so it's our time to shine. I'm doing a 10 star run right now. I got a question tho. How does the mouse handle the boat.

> Yep, there has been a boom in people talking and recommending our project since the announcement. It's exploded a bit on social platforms.

**Comment by Polymega** [comment]
@MasterDarkseid 

> Edit: The lastest build broke the camera rotation with the mouse he doesn't work. @mercury501

I tested the latest build and search camera still works with mouse. [Make sure you're using the latest build here.](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/618#issuecomment-1285040246)

**Comment by Polymega** [comment]
> Edit : nevermind he works now for whatever reasons

No worries. You might have had Search View set to "toggle" instead of "press" in the Game Options menu. If it's set to toggle and you activate it, Search View will be permanently enabled, which will always keep the camera behind James in free camera areas, until you toggle it again.

(I'm honestly not sure why there's an option to set this to "toggle" in the first place...)

![Screenshot 2022-10-20 21-21-16](https://user-images.githubusercontent.com/27395260/197089255-75d840dc-906b-4e25-99ca-cb9f32837f2e.png)

**Comment by MasterDarkseid** [comment]
So i guess i fugured out it's a bug that is random. My camera is now working but the mouse wheel for weapon cycle don't work.

So it's a bug that fix itself or not. When i started playing the camera was working at all. now the camera work but the wheel don't.

I guess it's random can you look into it.

**Comment by Polymega** [comment]
Got it! I've messaged Mercury about it. I found a peculiar bug myself:

When I was testing the toggle walk/run feature, I started a new game and noticed that pressing up on a controller's D-pad or left joystick wouldn't work. All other directions would make James move, but not up. Then, I loaded a new save file and it fixed itself. Through more testing, this bug only seemed to have happened when I started a new game from the main menu. How oddly specific/weird.....?

**Comment by MasterDarkseid** [comment]
Best way to test it it's to start a new game in the maria scenario. As the first two weapon in next to each other. Making it a fast way to see what cause this and if he works.

**Comment by Balthazor44** [comment]
I just started trying out the latest build, and I also have the mouse wheel no longer working, and additionally the run/guard no longer works either. I tried remapping those functions but that didn't work. And it's happening both on a new game and a loaded game save.

**Comment by mercury501** [comment]
I'll look into the random "up doesn't work" asap, regarding the wheel I've moved it to a separate config value and it's default to false, that's why it doesn't work. I'll write a little changelog next time next to the build!

**Comment by mercury501** [comment]
Hi all, here's a new build for you:
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9841703/d3d8.zip)

Patch notes:
- Implemented a new toggle walk/run feature found in Game Options > Movement.
   - The "analog" option has been changed to "toggle." With this, you no longer have to hold the walk/run button to change speeds on a keyboard. Simply press the button to toggle between the two. Full analog controller support is still supported when this is set to "toggle."
      - Note: Toggle only works with the keyboard, as analog is used for movement with a controller in this mode.
- Fixed an issue where, when run was toggled on for the new toggle walk/run feature, James wouldn't "strafe run" (he would "strafe walk")
- Fixed an issue where up/down on the D-pad and joystick of a controller sometimes wasn't responsive.
- Mouse wheel weapon swapping functionality has been moved to its own setting

**Comment by Balthazor44** [comment]
Cool! I should be able to spend a good deal of time testing tonight.

**Comment by mercury501** [comment]
You're dong so much good work @Balthazor44 , thank you for this.

**Comment by Polymega** [comment]
If you want to use/test the mousewheel for weapon swapping, be sure to enable it in the launcher or set `EnableMouseWheelSwap = 1` in the `d3d8.ini` file. 👍

**Comment by mercury501** [comment]
Hi @MasterDarkseid , routine checks, is `EnableEnhancedMouse` set to 1, and is your control type set to Rotational?

**Comment by Balthazor44** [comment]
I'm trying the latest build now, and I'm seeing that the run only works when the setting is set to "Toggle". If I set it to "Walk" or "Run" then pressing the run button does nothing. The "Toggle" setting seems to work great though.

**Comment by Polymega** [comment]
Great catch!

@mercury501 
To expand on Balthazor's findings: "Toggle" for movement type is working as intended, but it has an ill effect on "Walk" and "Run" movement types.

Let's say you have LCtrl set for "Run/Guard" in the Control Options menu. If you assign movement type to "Run" (where James will run by default), holding LCtrl should make him walk for as long as you have it held down (and vice versa if you set movement type to "Walk"). However, this is currently not happening.

**Comment by Balthazor44** [comment]
Here's a weird one, when I row across the lake, once I get out of the boat it's as if the up arrow or "move forward" button is stuck. James will immediately start walking with no button pressed. Going into a menu will shoot the selection up to the top. And there seems to be no way to "un-stick" it outside of restarting the game. I've tried it with the movement set to "Walk" and "Toggle" and it's the same either way.

**Comment by Polymega** [comment]
Very odd. I tested it a couple times myself and never had this happen. Is this consistently/always happening for you?

**Comment by Balthazor44** [comment]
That is odd. I'll try it on my other machine and see what happens there. I had it happen consistently four times in a row. I load the shipping dock game save, I hit Esc to exit the boarding animation, I try to steer the boat while holding down W to go forward, then his Esc again to exit the de-boarding animation. Then as soon as the game resumes James takes off running. I've tried it with the movement mode on toggle and walk, and I've crossed the lake pressing the run button periodically and without pressing it at all. All four times it was the same.

**Comment by Balthazor44** [comment]
It does the exact same thing on my other machine. I am using the same keyboard & mouse on each machine, so there is that in common. I'm going to experiment a bit, I'm curious to see if there's a way for me to have it not happen, like not skipping the animations.

**Comment by Polymega** [comment]
Yeah, I'm wondering if you're pressing a certain key combination or doing a certain action to provoke that.

As an FYI, Mercury had to do some special coding and special exemptions to make mouse movement with the boat work. So there might be some funny business for that reason.

**Comment by Balthazor44** [comment]
So far it happens no matter what I do. I tried not skipping the animations. I tried not even touching the mouse and just using the keyboard to navigate. I've tried holding down the forward key all the way, and tried releasing the forward key periodically. I'm going to try a few more things.

**Comment by Balthazor44** [comment]
When I'm holding down the forward key all the way until the "arrival" cutscene and release it after that's started, then the forward key will "stick,", but if I release the forward key before triggering the cutscene and just let the boat drift until the cutscene starts, then the problem doesn't occur. So, at least in my case, it seems to have something to do with having the forward key pressed when the cutscene is triggered.

**Comment by MasterDarkseid** [comment]
> Hi @MasterDarkseid , routine checks, is `EnableEnhancedMouse` set to 1, and is your control type set to Rotational?

Yes everything is set strange bug.

**Comment by mercury501** [comment]
Hi @Polymega and @Balthazor44 ,
here's a new build for you. 
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9844175/d3d8.zip)

I fixed the sprint function when something other than "Toggle" is selected, and should have fixed the input sticking after the boat stage. Tested it once, holding down W when the cutscene started, and didn't stick.

**Comment by Balthazor44** [comment]
I'll try that out right away. Did anyone besides me manage to duplicate that issue?

**Comment by MasterDarkseid** [comment]
> Hi @Polymega and @Balthazor44 , here's a new build for you. [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9844175/d3d8.zip)
> 
> I fixed the sprint function when something other than "Toggle" is selected, and should have fixed the input sticking after the boat stage. Tested it once, holding down W when the cutscene started, and didn't stick.
> 
> @MasterDarkseid could you provide the d3d8.log right after a session where you can't use the mouse to turn?

[d3d8.log](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9844825/d3d8.log)

**Comment by Balthazor44** [comment]
I just rowed across the lake a few times and it's working great now. Also the run button is back to working as well. Great work! Those were the only two problems I had with that build, everything else seemed to be working great. I'll do more testing later and see how that goes.

**Comment by mercury501** [comment]
@Balthazor44 that's great to hear! thank you.
@MasterDarkseid I don't see anything weird in the log, but I think I know what's happening. You seem to have  controller connected, could be that the controller's left analog isn't perfectly centered, and if there's input from the controller the mouse is ignored. Try this build, where I've implemented some tolerance on the left analog stick:
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9844875/d3d8.zip)

**Comment by Polymega** [comment]
Piggy-backing off of Mercury's post here, he just implemented another excellent--and what should be the final!--feature for mouse support: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9845158/d3d8.zip)

^ In this build, you can now fully control the search camera with the mouse (when activating search cam) while James is moving! Before, this wasn't possible. With this change, mouse support is identical to SH3 PC now. 😊 

------------------------------

To further explain this:

**For SH3 PC:**
When search cam is activated, search cam is given exclusive access to all mouse movement. Mouse movement does not control/influence Heather's movement when search cam is activated. Therefore, only keyboard inputs will move Heather when search cam is activated.

When search cam is not activated, left/right mouse movement will turn Heather like normal (in 3D movement mode).

**There were two issues with SH2 PC:**
1) Playing on keyboard, if James was not moving before activating search view, he will never move until you deactivate search view.
2) Playing on keyboard, if James is already moving when activating search view, mouse movement has no influence/control for search cam movement. You must stop James to have the mouse control the search cam, and this also brings you back to the first issue mentioned right above. 

...now, these two issues are fixed and it behaves like SH3 PC. 😊

**Comment by MasterDarkseid** [comment]
> @Balthazor44 that's great to hear! thank you. @MasterDarkseid I don't see anything weird in the log, but I think I know what's happening. You seem to have controller connected, could be that the controller's left analog isn't perfectly centered, and if there's input from the controller the mouse is ignored. Try this build, where I've implemented some tolerance on the left analog stick: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9844875/d3d8.zip)

The patch haven't fixed it however, you are right when i disconnected the controller the mouse camera work again. So if i want to play mouse i have to unplug my controller he seems.

**Comment by Polymega** [comment]
> The patch haven't fixed it however, you are right when i disconnected the controller the mouse camera work again. So if i want to play mouse i have to unplug my controller he seems.

Yeah, if your controller has a ton of "joystick drift," such as caused through years of heavy use, it could be _slightly_ sending Movement signals. With Mercury's mouse feature, controller movement will take precedence over mouse movement.

**Comment by MasterDarkseid** [comment]
> > The patch haven't fixed it however, you are right when i disconnected the controller the mouse camera work again. So if i want to play mouse i have to unplug my controller he seems.
> 
> Yeah, if your controller has a ton of "joystick drift," such as caused through years of heavy use, it could be _slightly_ sending movement signals. With Mercury's mouse feature, controller movement will take precedence over mouse movement.

My controller is a dualsense she had 1 year in september. But yeah controller as such low quality nowadays so i'm not surprised.

**Comment by MasterDarkseid** [comment]
I'm currently in a playtest right now. And notice something might be a bug.

While suing the hyper spray. He seems that mouse controls can sometime bug and what happen is the aim key (Right mouse button) sometime act both a aim and action.

Can someone replicate this.

**Comment by Polymega** [comment]
I just tried and couldn't make that happen.

Just to be sure: On Beginner (and Easy?) Action difficulty, James will immediately begin to attack if he's near an enemy and you press Action/Confirm (left click). For the other Action difficulties, you must first Ready Weapon (right click) then attack with Action/Confirm (left click). Could this be what's happening for you?

**Comment by mercury501** [comment]
The thing is, in theory, this should be impossible with my implementation. RMB can only behave as Cancel or Ready weapon, must be something external like a controller connected, non standard keybinds and the like.

**Comment by Balthazor44** [comment]
I have been doing testing on the latest build this evening, I just noticed that when movement is set to Toggle, then pressing the search camera button also makes James run. I don't see this happening in Walk or Run mode. It's great though to be able to use the search camera while James is moving now. Other than that, everything else so far has been working great.

**Comment by Polymega** [comment]
You've been on a roll catching these quirks, Balthazor! Great work and thanks!

@mercury501 
Some more specifics about this: When walk/run is set to "toggle" _and James is currently toggled to walk_, he will run when search cam is activated. 

When being looked into, you'll just want to make sure he will still run when "toggle" is set to run while search cam is activated.

**Comment by mercury501** [comment]
Hi all,
@MasterDarkseid I'll try and reproduce it, but that sounds really strange. Could be more of a game bug than a my bug.
@Balthazor44 Here's a patch, good catch on that. 
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9868125/d3d8.zip)

Since I inject left analog stick when using the search view, on analog run option, a full tilt is running. So, right now, when you're in toggle sprint mode and use the search view, it switches control option for that instant, and switches back when you stop using search view. 
It should be tested with a controller tho, to see if it interfered with something else, I'll try later if I manage.

**Comment by Balthazor44** [comment]
I did quite a bit of testing the past few days, and I'm not seeing any other issues. The search view works great now! I also played through with a controller again and had no issues there either. Seems pretty solid now.

**Comment by Balthazor44** [comment]
There is one thing I noticed, though I don't think it has anything to do with these mouse support changes because it seemed to work this way already. When using a controller with the control type set to Directional, if I use the search view to look around while James is moving, he will turn to move in the direction he's looking. I checked the PS2 version to see what that does, and it doesn't work that way. James keeps walking straight while looking around in the search view. And it only does this with the Directional control type, the Rotational works fine. For all I know, maybe the PC version always worked this way, but I figured I'd mention it.

**Comment by Polymega** [comment]
What I do as a quick and lazy solution: You can temporarily remove the `d3d8.dll` file to more-or-less play the PC game in vanilla.

**Comment by PedroC47** [comment]
hi everyone sorry for not being in touch in a while but my mobo died and i was saving for a new pc... i got a new one so i'll try to test it tomorrow if i have time. 
thank you :)

**Comment by PedroC47** [comment]
so far nothing strange to report:
-mouse inputs (confirm/cancel) work fine, attacking too, 
-search cam works too, using alt-tab and other things like that didn't break anything so far
my only improvement suggestion so far would be a way to use the mouse to navigate the map (left-click to drag the map "panning?") i don't know if this is hard to implement

**Comment by Polymega** [comment]
The game has native mouse support for the maps, but it isn't click + drag. When zoomed in on a map, click will move the map to where you clicked. Right click will zoom out of the map and you can then click the purple arrows to cycle between maps.

**Comment by MasterDarkseid** [comment]
So honestly this test is not a hassle like the audio bug, This one is smooth he seems, personally did like 3 or 4 full ran including the maria scenario and it's cool. Except the bug i've mentionned above when using the spray and during specific input combination. The build is bug-free to me.

**Comment by MasterDarkseid** [comment]
When will we be able to test the 60 FPS patch, if there's one planned.

**Comment by Polymega** [comment]
@Balthazor44 

> There is one thing I noticed, though I don't think it has anything to do with these mouse support changes because it seemed to work this way already. When using a controller with the control type set to Directional, if I use the search view to look around while James is moving, he will turn to move in the direction he's looking. I checked the PS2 version to see what that does, and it doesn't work that way. James keeps walking straight while looking around in the search view. And it only does this with the Directional control type, the Rotational works fine. For all I know, maybe the PC version always worked this way, but I figured I'd mention it.

I spoke with Mercury about this and he said he'll take a look into it! So to recap:

**When using 2D control mode...**
- On a controller, search cam movement should have no influence over James' movement, the same as it behaves in SH3 PC.
   - Most obvious in large, open areas where the camera can freely follow James.
- On keyboard, James currently comes to a dead stop when activating search cam. Have it to where he can still move while using search cam, with the same behavior as what's desired for controller usage (directly above).

**Comment by Polymega** [comment]
@MasterDarkseid 

> When will we be able to test the 60 FPS patch, if there's one planned.

There's still some outstanding high framerate issues we'd like to address first. Once most/everything is addressed, we can certainly do another large-effort for testing, the same as here for mouse support testing!

**Comment by matheusvb3** [comment]
Wish I could have helped with the mouse testing but my computer's PSU got toasted, pretty sure I'll be able to help with testing the 60FPS patch once it gets there though

**Comment by mercury501** [comment]
Hi all! Sorry for the long hiatus. 
Here is a new build, should fix the 2d controls search view.
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9997933/d3d8.zip)

**Comment by Polymega** [comment]
Thanks, Mercury! With a quick test I noticed James no longer comes to a dead stop when activating search cam in 2D mode while using mouse + keyboard, but he is now coming to a dead stop when using a controller. (Basically, the situation is now reversed!)

**Comment by Balthazor44** [comment]
Hopefully I should have some time tonight to take it for a spin. If not tonight then tomorrow for sure.

**Comment by mercury501** [comment]
You are absolutely right, here's a fix, I tried with both keyboard and controller, 2d and 3d and I *think* it works as it should.
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10002612/d3d8.zip)

**Comment by Balthazor44** [comment]
I did quite a bit of testing the past couple days with the most recent build, both with the mouse and controller. These are the only things I noticed:

1) when using the mouse/keyboard, switching to the search view while James is running causes him to momentarily slow to a walk then resume running. This doesn't seem to happen with the controller. Personally, this doesn't seem like much of an issue to me, and I suspect most people wouldn't even notice, but I figured I'd point it out.

2) when using a controller, I am still seeing the issue where James will move in the direction he is looking while in search view. This is with the control type set to Directional.

**Comment by Polymega** [comment]
> 1. when using the mouse/keyboard, switching to the search view while James is running causes him to momentarily slow to a walk then resume running. This doesn't seem to happen with the controller. 

I wasn't able to replicate this _solely_ using a keyboard/mouse or _solely_ using a controller. I was able to replicate this by activating search view, moving James with either the controller or keyboard, then switching to moving him with the keyboard or controller. If this is what you're describing...

> Personally, this doesn't seem like much of an issue to me, and I suspect most people wouldn't even notice, but I figured I'd point it out.

Then yes, I agree this is a very specific, rare-to-happen situation and to not worry about it.

> 2. when using a controller, I am still seeing the issue where James will move in the direction he is looking while in search view. This is with the control type set to Directional.

I went back to see how SH3 handles this particular situation and it looks like there is unique, special code made to handle this. I'm not sure how easy/viable it would be to do something like this (or if it's worth the time investment), so I leave the answer to this in Mercury's court!

**Comment by Balthazor44** [comment]
>I wasn't able to replicate this solely using a keyboard/mouse or solely using a controller. I was able to replicate this by activating search view, moving James with either the controller or keyboard, then switching to moving him with the keyboard or controller. If this is what you're describing...

I went back and checked, and it only happens for me when I have the movement set to "Toggle". Setting it to "Walk" or "Run" works fine. But again, it seems like a minor little thing that I personally wouldn't worry about.

**Comment by mercury501** [comment]
Thank you for your incredible work you're doing testing this feature, it would've been so much harder without you all.

I tried to fix the momentary slowdown when entering search view @Balthazor44 found, I tested with directional and rotational and seem to work. Since the portion of code used is the same as the boat stage fix we implemented, can you test that as well to see if it still works as intended?  I tried a bit and seems to work.

Here's the build:
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10023243/d3d8.zip)

**Comment by Balthazor44** [comment]
I tried this out and the momentary slowdown is gone and the boat stage still works fine, however I stumbled across something else now, where switching the movement from "Toggle" to "Walk" or "Run" causes issues. To duplicate this, start a game and go into the options to set the movement to "Toggle" then return to the game. Hit the run button to toggle James into sprint mode, then go back into the options and set the movement to "Walk" and return to the game. Now when I hit the search view button it locks James into sprint mode. The opposite seems to happen when I switch the movement from "Toggle" to "Run", where now it locks James into walk mode.

**Comment by mercury501** [comment]
Oopsie, that makes sense yes. I'll fix it asap and post a new build, thank you SO much @Balthazor44 !

**Comment by mercury501** [comment]
Just a wrong parenthesis in a condition xD
Here's a fix:
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10026389/d3d8.zip)

**Comment by Balthazor44** [comment]
I know how that goes. I'll try it out a little later tonight.

**Comment by matheusvb3** [comment]
I played the Born From a Wish scenario with Wine 7.21 on Linux and the game crashed after I left Heaven's Night. Wine generated a backtrace which I can upload should it be helpful, skimming through it the crash was caused by Wine3D3, I have no clue as to why. SH2 has never crashed to me before on Linux. I restarted the game and everything (well, except for one thing, which I will mention shortly) went well, but using the scroll wheel to change weapons wasn't working for some reason. Also I'm not sure if this is because of Wine or if it's not even within the scope of this project, but does the scroll wheel work for selecting between options (like pickup confirmations and the "How do you want to insert the board?" prompt in the fireplace puzzle)? I thought I could choose by moving the mouse like in the pause menu, but apparently the only way to move the selection is with the keyboard. I also noticed that the game stuttered a tiny bit more when out in the streets and my computer seemed to be struggling just a little more to run the game (fan noise was a little louder than normal), I also have no idea why. It's probably a Wine thing.

I also wanted to say that it was very cool to play with the mouse, I thought it would take a lot of getting used to but I found it even easier than using the keyboard. Oh, and @mercury501, you rock! You did an excellent job with this.

I will play the main scenario as soon as I can and will report back.

**Comment by Polymega** [comment]
> but using the scroll wheel to change weapons wasn't working for some reason.

This is disabled by default. Be sure `EnableMouseWheelSwap = 1` is included in the INI file. I requested Mercury have this disabled by default because, if you spam the mouse wheel, it makes James look really goofy while moving. I was afraid people would complain this "ruins the immersion" but, now that I've sat on it for a while, the player should be responsible to not spam it if they're that worried about it. With that said, @mercury501, may we make weapon swapping enabled by default again?

> Also I'm not sure if this is because of Wine or if it's not even within the scope of this project, but does the scroll wheel work for selecting between options (like pickup confirmations and the "How do you want to insert the board?" prompt in the fireplace puzzle)? I thought I could choose by moving the mouse like in the pause menu, but apparently the only way to move the selection is with the keyboard.

That's not a bad idea. If there is a reliable address we can find that indicates when such a selection is currently being displayed, then it should be no issue to include such as feature since Mercury has already done all the heavy lifting with researching mouse wheel/input movement.

> I also wanted to say that it was very cool to play with the mouse, I thought it would take a lot of getting used to but I found it even easier than using the keyboard. 

Indeed! It was so awkward for me to use the mouse back in the day with SH3 PC. But now? I can't imagine using a keyboard without one for the game(s). It's so nice to have this option in SH2 PC now. You've really outdone yourself here, Mercury! 😊

**Comment by Polymega** [comment]
> Just a wrong parenthesis in a condition xD Here's a fix: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10026389/d3d8.zip)

@mercury501, unfortunately, I still experience that issue with search cam view. Set run/walk to "toggle" and have James toggled to run. When using a keyboard to activate search view, James will walk, not run.

**Comment by matheusvb3** [comment]
> This is disabled by default. Be sure `EnableMouseWheelSwap = 1` is included in the INI file.

Now it works, thanks, but only for mouse wheel down. If you have only one weapon and try to swap with the mouse wheel it will cause James to lock up momentarily, I believe if you use the keyboard to swap it only does this if you have more than one weapon, right?

> If there is a reliable address we can find that indicates when such a selection is currently being displayed, then it should be no issue to include such as feature since Mercury has already done all the heavy lifting with researching mouse wheel/input movement.

Nice, I wasn't sure if that was already implemented or not. Another good use for the scroll wheel would be in the control bindings menu. When you bring it up the first option is positioned at the center of the list, which causes the rest of the bindings to be off-screen. I see what they were going for (Gimp, the image manipulation program for Linux, also does this because the GTK developers like to adhere to very old user interface paradigms and they think it's easier to select what you want that way, instead of doing what basically every other program does nowadays, but I digress), though it's still weird and non-intuitive in my opinion. Also if you really want to extrapolate, maybe the apartment clock puzzle as well could also afford to use the scroll wheel? 😅😂

**Comment by mercury501** [comment]
You are correct, the fix was completely wrong. Now it should be working as intended.
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10044825/d3d8.zip)

As for the scroll wheel puzzles it could be done yes, it surely would need a healthy dose of checks to make it work (@Polymega and I know something about that xD) but it's doable, tomorrow I'll have a look into finding some addresses.

**Comment by matheusvb3** [comment]
I tested on Linux up until the Abstract Daddy boss and I had two very tiny problems:

- I couldn't swap my weapon in the elevator trick or treat ride. Not important at all but I noticed it because the game forcefully swapped my chainsaw for the shotgun when I entered the elevator (it happened only in the trick or treat elevator ride, no idea why honestly)
- When I went to the storage room to get the items from the trick or treat box, I saved just before I opened it, failed the combination, reloaded, and went to answer again, for some reason the cursor was moving by itself to the left. It was pretty funny trying to hit the combination like this

Besides that, everything worked perfectly. I **think** the game was struggling a little more than normal when out in the streets, but this is just guesswork.

> As for the scroll wheel puzzles it could be done yes, it surely would need a healthy dose of checks to make it work (@Polymega and I know something about that xD) but it's doable, tomorrow I'll have a look into finding some addresses.

I was just joking! If you do that then it would open a Pandora's box of situations where the mouse wheel could be incorporated. Just off the top of my head I can think of the safe in the apartments and the combination locks in the hotel suitcase and hospital box. It's really not important and I imagine it would be a huge waste of development time.

Also I was curious about the weapon swap spazzing that Polymega mentioned, is it possible to trigger the weapon swap only if the wheel has travelled a certain delta? I think that could prevent it.

**Comment by mercury501** [comment]
Uhm I'll try and replicate the second point there, sounds like something really annoying. Did you have to restart the game to fix it?
if you try and replicate it, could you screenshot the debug overlay for me? (CTRL + G).

Thing is, the scroll wheel isn't actually analog to my knowledge, it just ticks in discrete increments, so I could like make it take 2 ticks to swap but I think that'd be more confusing than anything.

**Comment by matheusvb3** [comment]
> Did you have to restart the game to fix it? if you try and replicate it, could you screenshot the debug overlay for me? (CTRL + G).

No, I didn't have to restart the game. I failed the combination again and got hit with the acid vapor trap so I reloaded the save and that solved it. I will try to get it to get it to happen again.

> Thing is, the scroll wheel isn't actually analog to my knowledge, it just ticks in discrete increments, so I could like make it take 2 ticks to swap but I think that'd be more confusing than anything.

Oh, bummer. I always thought the mouse delta was analog for some reason.

> The mouse wheel weapon swap is reliant on the keyboard inputs that would normally do this. I checked and weapon swapping is disabled during this elevator radio show. So if it's disabled for keyboard, it'll also be disabled for the mouse. For all intents and purposes, this is "normal" behavior.

That's weird, but like I said it's not important. I only noticed it because James was holding the chainsaw before entering the elevator and when the screen transitioned to him inside he was holding the shotgun for some reason.

> I tried replicating this but couldn't. Were you able to consistently make this happen? Did you have a controller plugged in that you maybe accidentally touched and the joystick started to drift?

No, I was only using the keyboard and mouse. I will try to get it to happen again.

> That might be a placebo effect. The mouse feature shouldn't affect anything like that in the game.

I think that might be a placebo too, but Wine has some really weird quirks. I will try to run around the town without the mouse control and see if it still does that.

> Yeah. We just need to have faith a player won't be constantly spamming the mouse wheel just because they can.

To be fair, it's really funny at first! Same with making James headbang with the search camera when you're in a spot where the camera is completely fixed.

**Comment by matheusvb3** [comment]
I got the bug to happen again, this time with the padlock to the third floor of the hospital. I recorded the screen with Kazam, an application for screen recording on Linux, but unfortunately it captured only a few frames of the game.

https://user-images.githubusercontent.com/98937378/202837236-1d468c99-9de9-4e45-bd27-c2546087f489.mp4

It has the debug info overlay enabled. I try to hit the correct combination on the padlock, exit out because it was too frustrating, then when I went to try again the mouse went back to working correctly. After that I shot Maria a few times and chainsawed her for no real reason.

I used the saves given at #28 to get to the hospital quicker.

I will try to get it to happen again and will use a better program to record the screen.

**Comment by mercury501** [comment]
Ok, I have a theory, I think it happens when, interacting with the puzzle, the mouse is moving horizontally. Can't test it right now, but I think that since you're in a puzzle it doesn't poll the mouse anymore, so the last known value is used. I'll test it later

**Comment by mercury501** [comment]
This is a tentative fix:
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10046994/d3d8.zip)
basically now, if you're in a full screen image event, it zeroes out the mouse x axis. Should not interfere with controller or with events that use the mouse cursor.

**Comment by matheusvb3** [comment]
> Ok, I have a theory, I think it happens when, interacting with the puzzle, the mouse is moving horizontally.

That's correct, I just checked and the cursor moves by itself if you were moving the mouse horizontally before a puzzle screen comes up. It can happen with any screen that requires the cursor to interact with the puzzle. For some reason it only happens with the mouse x-axis, the y-axis doesn't result in any strange behavior.

> This is a tentative fix: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10046994/d3d8.zip) basically now, if you're in a full screen image event, it zeroes out the mouse x axis. Should not interfere with controller or with events that use the mouse cursor.

It works perfectly. Thanks!

Just to be sure, is this a Wine-only thing or can it be reproduced in Windows? If so I could file a bug to Wine. I've already had Wine act weird with the mouse in another game, Max Payne ([you can check the bug report here](https://bugs.winehq.org/show_bug.cgi?id=53627)), and the developers were able to bisect what caused it to start acting weird.

By the way I'll upload the backtrace I mentioned in my previous comment in case it could be important, it was apparently caused by WineD3D so it shouldn't be due to the mouse controls but you never know:

[backtrace.txt](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10047707/backtrace.txt)

**Comment by matheusvb3** [comment]
@Polymega Now I know why only the wheel down to swap weapons was working. It was because I had rebinded E to be the search camera key. Now I set run to shift and search cam to ctrl like in option B in the image of your post and it works correctly.

**Comment by mercury501** [comment]
@matheusvb3 that's great! No it's not a wine thing, I reproduced it in windows no problem, it's a sh2 thing.

As for the overlay it's not imgui, it's pure directx 8, just renders a bit of text. If you're courious about the code, it's [here](https://github.com/mercury501/Silent-Hill-2-Enhancements/blob/Input-Test/Wrappers/d3d8/Overlay.cpp)

I'm doing some experiments with ImGui, but haven't merged them yet. if you're courious, here's a simple test (very early, I'm still thinking about how to make it work) , it's enabled with CTRL + F
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10047836/d3d8.zip)

**Comment by matheusvb3** [comment]
> As for the overlay it's not imgui, it's pure directx 8, just renders a bit of text.

Oh, alright, I have no experience whatsoever with graphical computing so I had no idea you could implement debug overlays with DirectX.

I tested the build you linked and yes it's buggy, bug I imagine the possibilities for modding would become even greater when it's finished (along with the level editor which is being worked on. Frankly I think there's a huge untapped potential for SH-like experiences in gaming, the only game to scratch that itch for me was Cry of Fear, but I'm just rambling...)

@Polymega I would like to ask how do you select when to use heavy or light attacks with melee weapons on PC? My understanding from playing the game and from the last update video on Youtube is that the first attack should always be strong and if you keep the attack button held down the next attacks will be light ones, but I was playing on hard when testing mouse controls and sometimes James will do a light attack first for no reason when I intended to do a heavy one. It's a little annoying because the light chainsaw attack is a slow horizontal cut that often ends up hitting the walls and cancelling out, and just now I went to fight Eddie with the pipe cheese but James insists on doing the light poke attack for some reason.

**Comment by Polymega** [comment]
Here's some basic info on that: 
https://enhanced.townofsilenthill.com/SH2/troubleshoot.htm#how-do-i-switch-between-light-and-heavy-melee-attacks

> but I was playing on hard when testing mouse controls and sometimes James will do a light attack first for no reason when I intended to do a heavy one

I think that might be happening if you were moving when you were trying to initiate a heavy attack (by holding down the attack button). 

The game was clearly made around pressure sensitive buttons, that's for sure.
>

**Comment by mercury501** [comment]
Hi all!
Thank you for your contribution to bringing this feature to life.
I wanted to ping you again to ask you if you had any "last minute" pointers and issues that you find with the latest build, just in case, don't hesitate to post them !

**Comment by Balthazor44** [comment]
I haven't had much of a chance to do much lately, but I'll spend at least a few hours tonight taking the latest build for a spin and let you know later how it goes.

**Comment by Balthazor44** [comment]
@mercury501 when you get a chance can you link the latest build again? I'm still seeing issues with the "Toggle" movement setting so I'd like to make sure I'm using the correct build.

**Comment by mercury501** [comment]
hi @Balthazor44 it should be the one in this post [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/618#issuecomment-1320956564), when I get home I'll build again and upload it!

**Comment by matheusvb3** [comment]
I won't be able to test unfortunately, I'm busy with some stuff at the moment. :(

Perhaps around the holidays I'll be able to do a full and thorough playthrough as I'll have a short break from uni, maybe by then 60FPS will also be ready for testing? Wishful thinking!

**Comment by mercury501** [comment]
Here's the latest build:
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10103731/d3d8.zip)

**Comment by Balthazor44** [comment]
Ok great, I will take that for a spin this evening.

**Comment by Balthazor44** [comment]
I just tried the latest build and I'm still seeing an issue with the toggle movement mode. The first time I hit the look view button James will behave as though it is set to the toggle mode, and there is no way to fix it short of restarting the game. Aside from that, everything I've tried works great.

Was there anything done to fix the problem of James turning in the direction he's looking when using the look view with a controller? I know there was discussion earlier about whether or not that was worth the effort, or even possible, so I'm curious where we landed on that.

**Comment by Polymega** [comment]
> The first time I hit the look view button James will behave as though it is set to the toggle mode, and there is no way to fix it short of restarting the game.

Can you expand upon this a bit more?

> Was there anything done to fix the problem of James turning in the direction he's looking when using the look view with a controller?

It hasn't been looked into any further since then. SH3 PC has some special code going on to handle this, and we're not quite sure how to replicate its behavior.

**Comment by Balthazor44** [comment]
Ok, for example if I start a game and I have the movement mode set to "Walk" James will function in walk mode until I hit the look view button. From that point on James will function like he's in "Toggle" mode. The options screen still shows me in walk mode, but it's working like toggle mode. And I have not been able to find a way to return it to functioning in walk mode. I tried changing the movement to toggle and back again, and I tried setting it to run mode. I've tried loading a game save. I've tried quitting to the main menu and starting a new game. So basically, as soon as I hit the look view button James gets permanently stuck in toggle mode.

**Comment by Polymega** [comment]
Thanks for going into more detail. I think I know why this started happening.

A few weeks ago I was also playtesting this for Mercury and found that, _while_ search cam is activated, pressing the toggle key did not toggle James' movement.

Mercury put in a fix for this but it looks to be affecting the non-toggle movement modes as well. It might just be something simple as not checking whether toggle mode is currently being used for this search cam check.

**Comment by Balthazor44** [comment]
Makes sense. I've gone through about half the main scenario now, and also played through the rowboat section, and tried everything else that's been an issue so far, and everything is working great outside of that toggle issue. While it would be ideal to get the controller look view issue resolved, it definitely doesn't seem to me like a big enough problem to spend enormous amounts of time and effort trying to fix.

**Comment by mercury501** [comment]
Ok, I'll look into it as soon as I can, should be an issue of when to activate the special function @Polymega talked about

**Comment by mercury501** [comment]
Hi all!
I think I fixed it, here's the build:
[d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/10116594/d3d8.zip)

tested with toggle and walk, and they look to behave as intended.

**Comment by Balthazor44** [comment]
I just tried out that build and it works for me too. I tried everything I could think of to get it stuck in toggle mode again and it worked fine every time. I'll probably mess around with it some more later, but so far so good.

**Comment by mercury501** [comment]
@Balthazor44 we are having issues replicating a bug reported in #694, do you have experience with the hyper spray by chance?

**Comment by Balthazor44** [comment]
> @Balthazor44 we are having issues replicating a bug reported in #694, do you have experience with the hyper spray by chance?

Yeah but I don't currently have a hyper spray, I'll need to finish the game again to get one. I can do that this weekend and try duplicating the issue.

**Comment by MasterDarkseid** [comment]
new issue find but it's unrelated to the patch. I think this is present on any version of SH2.

That issue is.... OBS making the game crash.

**Comment by Polymega** [comment]
Are you using "Window Capture" for the game with OBS? More info here: https://enhanced.townofsilenthill.com/SH2/troubleshoot.htm#what-is-the-best-way-to-streamcapture-the-game-using-obs

**Comment by MasterDarkseid** [comment]
> Are you using "Window Capture" for the game with OBS? More info here: https://enhanced.townofsilenthill.com/SH2/troubleshoot.htm#what-is-the-best-way-to-streamcapture-the-game-using-obs

I'm using screen capture. i'm not in 1440P due to my new monitor but even my old in 1080p was doing this. He might be OBS issue.

**Comment by MarioTainaka** [comment]
> > Are you using "Window Capture" for the game with OBS? More info here: https://enhanced.townofsilenthill.com/SH2/troubleshoot.htm#what-is-the-best-way-to-streamcapture-the-game-using-obs
> 
> I'm using screen capture. i'm not in 1440P due to my new monitor but even my old in 1080p was doing this. He might be OBS issue.

Are you running in Fullscreen Windowed mode?

**Comment by MasterDarkseid** [comment]
let me check

**Comment by MarioTainaka** [comment]
Then it should just work. try using Window Capture or game source.

**Comment by MasterDarkseid** [comment]
so i have installed the new update and this occur. is that normal

![image](https://user-images.githubusercontent.com/104478182/221296990-8a1e40cd-ce68-4db2-b83b-51fe38be2f50.png)

**Comment by MasterDarkseid** [comment]
![image](https://user-images.githubusercontent.com/104478182/221297291-f7ef811f-a5e2-4320-85e9-0579246d394d.png)

**Comment by MasterDarkseid** [comment]
Is that normal i've installed the patch using the setup

![image](https://user-images.githubusercontent.com/104478182/221299078-8057272a-2db1-436f-acbf-60c576e8d9c1.png)

**Comment by Polymega** [comment]
Did you also update the "Enhanced Edition Essential Files" package? The restored Japanese language files are contained in there.

**Comment by MasterDarkseid** [comment]
![image](https://user-images.githubusercontent.com/104478182/221300657-b1a65a0f-c421-4b98-affd-95700cfa4bbb.png)

i got this error now

**Comment by Polymega** [comment]
Yeah... Servers are wrecked from the traffic spike... again. Two hours later with customer service and I have nothing to show for it. 

Well, that's not true. After the first call, which lasted an hour, they closed my ticket a few hours later saying they found no issues on their end. Another call and another hour on the phone later, I expressively illustrated my thoughts and opinions on their lack of accountability and follow-through. 

However, I can only do so much over phone. I'm ultimately at their mercy with how quick they rectify this.

**Comment by MasterDarkseid** [comment]
Oh so it s a hoster issue.

**Comment by MasterDarkseid** [comment]
@Polymega so i did some checking and yeah everything is installed but the JP language is missing.

**Comment by Balthazor44** [comment]
I noticed with the update that I have no way to install the Xidi or Credits options. It says "Not Installed" in the setup tool and they can't be selected. If you'd like me to open a new ticket for this, I can, but I figured I'd try here first.

**Comment by Balthazor44** [comment]
Actually never mind, I'm just an idiot. I just now realized I have to go to the Install or Repair Packages screen first. My bad.

**Comment by MasterDarkseid** [comment]
The patch is awesome enhanced Edition team. You can't imagine how happy i am to be able to experiment my favorite game with such great improvement.

I wish every game was like this kind of love.

SH3 and SH4 and homecoming could really benefit from your expertise.
